File: app/src/main/java/xyz/zpayh/myadapter/HeadAndFootActivity.java
package xyz.zpayh.myadapter;

import android.os.Bundle;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import java.util.ArrayList;
import java.util.List;

import xyz.zpayh.adapter.DiffUtilCallback;
import xyz.zpayh.myadapter.adapter.HeadAndFootAdapter;
import xyz.zpayh.myadapter.data.ImageCard;

public class HeadAndFootActivity extends AppCompatActivity {

    private boolean mChangeOrder = true;

    private String mTitles[] = {"Adult","Easter Eggs","Girl", "Sunset"};
    private int mImageResId[] = {R.drawable.adult, R.drawable.easter_eggs, R.drawable.girl, R.drawable.sunset};

    private HeadAndFootAdapter mAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_head_and_foot);

        RecyclerView recyclerView = (RecyclerView) findViewById(R.id.list);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        mAdapter = new HeadAndFootAdapter(this);
        recyclerView.setAdapter(mAdapter);

        mAdapter.addHeadLayout(R.layout.item_head);
        mAdapter.addFootLayout(R.layout.item_foot);
        mAdapter.addFootLayout(R.layout.item_foot2);

        List<ImageCard> data = new ArrayList<>();

        for (int i = 0; i < mTitles.length; i++) {
            ImageCard card = new ImageCard(mImageResId[i],mTitles[i]);
            data.add(card);
        }

        mAdapter.setData(data);

        final SwipeRefreshLayout refresh = (SwipeRefreshLayout) findViewById(R.id.refresh);
        refresh.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                refresh();
                refresh.setRefreshing(false);
            }
        });

        findViewById(R.id.fab).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                refresh();
            }
        });
    }

    private void refresh() {
        if (mChangeOrder){
            List<ImageCard> data = new ArrayList<>();
            for (int i = mTitles.length - 1; i >= 0; i--) {
                ImageCard card = new ImageCard(mImageResId[i],mTitles[i]);
                data.add(card);
            }
            mAdapter.setData(data);
        }else{
            List<ImageCard> data = new ArrayList<>();
            for (int i = 0; i < mTitles.length; i++) {
                ImageCard card = new ImageCard(mImageResId[i],mTitles[i]);
                data.add(card);
            }

            mAdapter.setData(data);
        }
        mChangeOrder = !mChangeOrder;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_anim:
                mAdapter.setCallback(new DiffUtilCallback<ImageCard>() {
                    @Override
                    public boolean areItemsTheSame(ImageCard oldItem, ImageCard newItem) {
                        return oldItem.mImageResId == newItem.mImageResId;
                    }

                    @Override
                    public boolean areContentsTheSame(ImageCard oldItem, ImageCard newItem) {
                        return TextUtils.equals(oldItem.mImageTitle,newItem.mImageTitle);
                    }
                });
                return true;
            case R.id.action_not_anim:
                mAdapter.setCallback(null);
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.head_menu, menu);
        return true;
    }
}

File: sample-kotlin/src/main/java/cn/sherlockzp/sample/HeadAndFootActivity.kt
package cn.sherlockzp.sample

import android.os.Bundle
import android.app.Activity
import android.support.v7.app.AppCompatActivity
import android.support.v7.util.DiffUtil
import android.support.v7.widget.LinearLayoutManager
import android.view.Menu
import android.view.MenuItem
import android.widget.ImageView
import cn.sherlockzp.adapter.BaseAdapter
import cn.sherlockzp.adapter.BaseViewHolder
import cn.sherlockzp.adapter.ImageCallback
import cn.sherlockzp.adapter.LayoutConfig
import cn.sherlockzp.sample.model.ImageCard
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions

import kotlinx.android.synthetic.main.activity_head_and_foot.*

class HeadAndFootActivity : AppCompatActivity() {

    private var changeOrder = true

    private val titles = arrayOf("Adult","Easter Eggs","Girl", "Sunset")
    private val imageResIds = arrayOf(R.drawable.adult, R.drawable.easter_eggs, R.drawable.girl, R.drawable.sunset)

    private val callback = object : DiffUtil.ItemCallback<ImageCard>(){
        override fun areItemsTheSame(oldItem: ImageCard, newItem: ImageCard): Boolean {
            return oldItem.imageResId == newItem.imageResId
        }

        override fun areContentsTheSame(oldItem: ImageCard, newItem: ImageCard): Boolean {
            return oldItem.title == newItem.title
        }
    }

    private val adapter = object : BaseAdapter<ImageCard>(){
        override fun getLayoutRes(index: Int) = R.layout.item_image_card

        override fun convert(holder: BaseViewHolder, data: ImageCard, index: Int) {
            holder.setText(R.id.tv_image_title, data.title)
                    .setImage(R.id.iv_bg) {
                        Glide.with(this@HeadAndFootActivity)
                                .load(data.imageResId)
                                .apply(RequestOptions.fitCenterTransform())
                                .into(it)
                    }
        }

        override fun convertHead(holder: BaseViewHolder, headLayout: LayoutConfig) {
            holder.setText(R.id.tv_head,"这是头部啦")
        }

        override fun convertFoot(holder: BaseViewHolder, footLayout: LayoutConfig) {
            holder.setText(R.id.tv_foot, "这是尾部")
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_head_and_foot)

        rv_list_girls.layoutManager = LinearLayoutManager(this)
        rv_list_girls.adapter = adapter

        initAdapter()

        sr_refresh.setOnRefreshListener {
            refresh()
            sr_refresh.isRefreshing = false
        }
    }

    private fun refresh(){
        if (changeOrder){
            val data = ArrayList<ImageCard>()
            for (i in titles.indices) {
                val card = ImageCard(imageResIds[i], titles[i])
                data.add(card)
            }
            adapter.setData(data.reversed())
        } else {
            val data = ArrayList<ImageCard>()
            for (i in titles.indices) {
                val card = ImageCard(imageResIds[i], titles[i])
                data.add(card)
            }
            adapter.setData(data)
        }
        changeOrder = !changeOrder
    }

    private fun initAdapter() {
        val data = ArrayList<ImageCard>()
        for (i in titles.indices) {
            val card = ImageCard(imageResIds[i], titles[i])
            data.add(card)
        }

        with(adapter) {
            addHeadLayout(R.layout.item_head)
            addFootLayout(R.layout.item_foot)
            addFootLayout(R.layout.item_foot2)
            setData(data)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when(item.itemId) {
            R.id.action_anim -> adapter.setItemDiffCallback(callback)
            R.id.action_not_anim -> adapter.setItemDiffCallback(null)
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.head_menu, menu)
        return true
    }
}
